Add reference implementation of EstimatorV2

ikkoham opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Add EstiamtorV2

This PR includes base classes and StatevectorEstimator (reference implementation).

### Details and comments

This PR is based on the following RFCs:
https://github.com/Qiskit/RFCs/blob/master/0015-estimator-interface.md
https://github.com/Qiskit/RFCs/blob/master/0018-primitive-options-type.md
https://github.com/Qiskit/RFCs/blob/master/0017-base-primitive-unification.md

Items not included in this PR:
- Estimation for non-Pauli observables
- SamplerV2
- Deprecation of V1 and primitives.utils
- BindingsArray with str key

Note: I might do a force push. → done

ikkoham committed:
Add EstimatorV2

jyu00 commented on qiskit/primitives/base/estimator_task.py:
```diff
@@ -0,0 +1,92 @@
+    """
+
+    observables: ObservablesArray
+    parameter_values: BindingsArray
```
`parameter_values` is not a required field and hence needs a default.

ikkoham replied:
Which do you think is None or BindingsArray() nice as the default?

ikkoham replied:
I updated. `BindingsArray(shape=())` is the default value of `EstimatorTask.paramter_values`

jyu00 commented on qiskit/primitives/base/observables_array.py:
```diff
@@ -0,0 +1,247 @@
+
+    def __init__(
+        self,
+        observables: Union[BasisObservableLike, ArrayLike],
```
There was this comment in pec-runtime's `ObservablesArray`:

```
    # NOTE: With support for arrays of observables, its ambiguous how
    # to treat a non-sequence input, so we should deprecate passing in
    # a single BasisObservableLike
```
I'm not sure why single `BasisObservableLike` is ambiguous, but it'd need to be converted or `tolist()` won't return a list.

jyu00 commented on qiskit/primitives/base/estimator_task.py:
```diff
@@ -0,0 +1,92 @@
+        self.parameter_values.validate()
+        # Cross validate circuits and observables
+        for i, observable in enumerate(self.observables):
+            num_qubits = len(next(iter(observable)))
```
This validation doesn't work when passing in a `PauliList` for a circuit.

ikkoham replied:
PauliList will be deprecated https://github.com/Qiskit/qiskit/pull/11055

jyu00 commented on qiskit/primitives/base/estimator_task.py:
```diff
@@ -0,0 +1,92 @@
+    """
+
+    observables: ObservablesArray
+    parameter_values: BindingsArray
```
Perhaps we should enforce type checking, so people (like me) don't try to do `task = EstimatorTask(circuit, ["XX"])`.

ikkoham replied:
I updated to use `pydantic.dataclass` instead of `dataclasses.datacalss`.  It should validate the type.

jyu00 commented on qiskit/primitives/base/bindings_array.py:
```diff
@@ -0,0 +1,356 @@
+                bindings_array = bindings_array.reshape((1, -1))
+            bindings_array = cls(bindings_array)
+        elif isinstance(bindings_array, Mapping):
+            bindings_array = cls(kwargs=bindings_array)
```
```suggestion
            bindings_array = cls(kwvals=bindings_array)
```

jyu00 replied:
Should add a final `else` and raise `TypeError` for invalid type

ikkoham replied:
Thanks

ikkoham commented on qiskit/primitives/base/estimator_task.py:
```diff
@@ -0,0 +1,92 @@
+    """
+
+    observables: ObservablesArray
+    parameter_values: BindingsArray = BindingsArray(shape=())
```
Property name `bindings_array` might be better?

jyu00 replied:
I think `parameter_values` is more intuitive, especially for people don't just pass in an array instead of creating a `BindingsArray` explicitly.

t-imamichi commented on qiskit/primitives/base/estimator_task.py:
```diff
@@ -0,0 +1,92 @@
+
+
+"""
+Estiamtor Task class
```
```suggestion
Estimator Task class
```

ikkoham replied:
Thanks

t-imamichi commented on qiskit/primitives/base/estimator_task.py:
```diff
@@ -0,0 +1,92 @@
+            task: an object to be estimator task.
+
+        Returns:
+            A coerced estiamtor task.
```
```suggestion
            A coerced estimator task.
```

ikkoham replied:
🙇‍♂️

t-imamichi commented on test/python/primitives/test_estimatorv2.py:
```diff
@@ -0,0 +1,285 @@
+        np.testing.assert_allclose(result4[0].data["evs"], [1.55555728, -1.08766318])
+        np.testing.assert_allclose(result4[1].data["evs"], [0.17849238])
+
+    def test_estiamtor_run_no_params(self):
```
```suggestion
    def test_estimator_run_no_params(self):
```

ikkoham replied:
🙇‍♂️

t-imamichi commented:
Will you implement `DataBin` defined by https://github.com/Qiskit/RFCs/blob/master/0017-base-primitive-unification.md#databin?

t-imamichi commented:
How about moving bindings_array.py, ovservables_array, and shape.py to a new directory `qiskit.primitives.containers`?

t-imamichi commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -258,3 +260,42 @@ def parameters(self) -> tuple[ParameterView, ...]:
+            tasks = [tasks]
+        elif isinstance(tasks, tuple) and isinstance(tasks[0], QuantumCircuit):
+            tasks = [EstimatorTask.coerce(tasks)]
+        elif tasks is not EstimatorTask:
```
Do you intend something like `elif isintance(tasks, list)`? I don't get why you check `is not`.

ikkoham replied:
This is because it would be complex with tuple branches above, but it is true that it is better to check properly and raise an exception.

ikkoham replied:
updated

ikkoham commented:
> Will you implement DataBin defined by https://github.com/Qiskit/RFCs/blob/master/0017-base-primitive-unification.md#databin?

Sure. It's in the TODO list in the issue comment. I am waiting for internal discussions for DataBin.

t-imamichi commented:
Do you plan to deprecate public methods in qiskit.primitives.utils? I think they are not used when qiskit.algorithms and PrimtiivesV1 are removed.

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -258,3 +260,44 @@ def parameters(self) -> tuple[ParameterView, ...]:
+    def __init__(self, options: Optional[BasePrimitiveOptionsLike]):
+        super().__init__(options=options)
+
+    def run(self, tasks: EstimatorTaskLike | Iterable[EstimatorTaskLike]) -> T:
```
I believe this should be `Job[T]` instead of `T`. However, `ibm.providers.Job` isn't `Generic[T]` IIRC, so I think we either have to just put `Job` or update `Job` to be generic.

ikkoham replied:
OK, I changed `run` method returns Job instead of T.

ihincks commented on qiskit/primitives/containers/options.py:
```diff
@@ -0,0 +1,35 @@
+
+@mutable_dataclass
+class BasePrimitiveOptions(ABC):
+    """Base calss of options for primitives."""
```
```suggestion
    """Base class of options for primitives."""
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -258,3 +260,44 @@ def parameters(self) -> tuple[ParameterView, ...]:
+class BaseEstimatorV2(BasePrimitiveV2, Generic[T]):
+    """Estimator base class version 2.
+
+    Estimator estimates expectation values of quantum circuits and observables.
```
```suggestion
    An Estimator estimates expectation values of quantum circuits and observables.
```

ihincks commented on qiskit/primitives/containers/primitive_result.py:
```diff
@@ -0,0 +1,49 @@
+T = TypeVar("T", bound=TaskResult)
+
+
+class PrimitiveResult(Sequence[T]):
```
As @jakelishman pointed out, `Sequence` adds in some methods like `count` that we don't really want to pretend to support. I think we can just keep this class as `Generic[T]`, and have

```
__iter__(self) -> Iterable[T]
__len__(self) -> int
__getitem__(self, int) -> T
```

ihincks commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,369 @@
+            bindings_array: an object to be bindings array.
+
+        Returns:
+            A coerced bindings array.
```
```suggestion
            A new bindings array.
```

ihincks commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,369 @@
+
+    @classmethod
+    def coerce(cls, bindings_array: BindingsArrayLike) -> BindingsArray:
+        """Coerce BindingsArrayLike into BindingsArray
```
```suggestion
        """Coerce an input that is :class:`~BindingsArrayLike` into a new :class:`~BindingsArray`.
```

ihincks commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,369 @@
+        if isinstance(bindings_array, Sequence):
+            bindings_array = np.array(bindings_array)
+        if bindings_array is None:
+            bindings_array = cls([], shape=(1,))
```
reminder to make sure this is consistent.

ihincks commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,369 @@
+            bindings_array = cls([], shape=(1,))
+        elif isinstance(bindings_array, np.ndarray):
+            if bindings_array.ndim == 1:
+                bindings_array = bindings_array.reshape((1, -1))
```
Is this leading dimension of 1 necessary? It would be great, if possible, if all such shape logic could be restricted to construction.

ikkoham replied:
There are many use cases of passing in one dimension array that I include it for user convenience. However, this is the same as my first implementation of the run method, which could take a single task. We can force the user to pass more than a two-dimensional array. Which would be better?

ihincks replied:
Unless I'm misunderstanding something, I would expect that a 1D np.ndarray input would correspond to the desire to estimate a single number-like as the observable (assuming observables doesn't broadcast it further). As in, the estimated expectation value would be an ndarray with shape ().

In this case we'd want the BindingsArray to end up with shape (), which I think can be supported by copying the latest iteration of BindingsArray as implemented elsewhere by @t-imamichi and @samanthavbarron (I can send you the link). For example, I see this test in their PR:
```python
            ba = BindingsArray(np.array([1, 2, 3]))
            self.assertEqual(ba.shape, ())
```

ihincks commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,369 @@
+        """
+        return self.reshape(self.size)
+
+    def reshape(self, shape: Union[int, Iterable[int]]) -> BindingsArray:
```
Reminder to fold in changes to this signature from elsewhere. That is, accept `*shape: ShapeInput`.

ihincks commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,369 @@
+        """The non-keyword values of this array."""
+        return self._vals
+
+    def bind_at_idx(self, circuit: QuantumCircuit, idx: Tuple[int, ...]) -> QuantumCircuit:
```
I'd like to rename  the argument `idx` to `loc` here, for consistency with, eg., BitArray. And then rename the method to `bind`.

ikkoham replied:
updated

ihincks commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,369 @@
+        arr = np.empty(self.shape, dtype=object)
+        for idx in np.ndindex(self.shape):
+            arr[idx] = self.bind_at_idx(circuit, idx)
+        return arr
```
These two functions seem a bit too similar. Could we just have one?

ikkoham replied:
I did not use `bind_flat`, I remove it.

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -15,7 +15,97 @@
 =====================
-Overview of Estimator
+Overview of EstimatorV2
+=====================
```
```suggestion
=======================
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -15,7 +15,97 @@
+
+=====================
+Overview of EstimatorV1
 =====================
```
```suggestion
=======================
```

ikkoham commented:
> Do you plan to deprecate public methods in qiskit.primitives.utils? I think they are not used when qiskit.algorithms and PrimtiivesV1 are removed.

IMO, at the same time with PrimitiveV1, so 0.46?

chriseclectic commented on qiskit/primitives/containers/task_result.py:
```diff
@@ -0,0 +1,32 @@
+
+
+@frozen_dataclass
+class TaskResult:
```
General question: What is the advantage of using dataclasses for these containers instead of just making our own regular class containers? dataclasses are note the best for subclassing because the order of attributes cant be changed or modified, and if a base has a default field, a subclass cant add any non-default fields to it. I think it would be better long term to avoid using dataclasses for the tasks and results or any other class we intend to use with subclassing or where we are adding public methods to them.

ikkoham replied:
There are pros and cons for using dataclasses.
Pros
1. Data validation (Type hints can be enforced)
2. As for options, we are supposed to use dataclasses from the RFC, so they can be common.

Cons
1. There will certainly be overhead.

I don't understand the issues related to inheritance. For example,
```python
from qiskit.primitives.containers.dataclasses import frozen_dataclass

@frozen_dataclass
class A:
    name: str = "a"

@frozen_dataclass
class B(A):
    number: int = 1
    
    def print(self):
        print(self.name, self.number)

b = B(name="b", number=3.)
b.print() # outputs b 3
```
works well. Also, the first mentioned validation works as follows:
![image](https://github.com/Qiskit/qiskit/assets/6814928/2c5e553b-669d-4b53-9e8e-50891e63c313)


@jyu00 Do you have a comment? The data structure and serialization part is important on the runtime side?

jyu00 replied:
Using a dataclass is very useful for options, especially with pydantic, since it does validation and converts dictionary inputs to nested dataclasses. It's probably less important for output that since we control that data (v.s. input data from users). 

@chriseclectic Do you have a specific use case in mind?

ihincks replied:
These classes are simple enough that I don't think dataclasses are buying us too much.

chriseclectic replied:
My concern is that for anything intended as a non-trivial base class with subclassing, like tasks and task results, dataclasses seem an inconvenient fit. Also anything that adds methods, or we want to use a custom repr.

With data classes you cant do something like:

```python
class A:
    def __init__(self, x, y=1)

class B(A):
    
    def __init__(self, a, x=2, y=3)
```
where you want to add a new positional argument. Or
```python
class B(A):
    def __init__(self, a=10):
        super().__init__(x=1, y=2)
```
where you want to fix values in base class.

You are stuck with the order of all attributes, and cannot add any positional ones, only kwvals (assuming the base class has any kwvals).

If you want to make a regular class act as "frozen" you just define a property and not a setter:

```python
def A:
     def __init__(self, arg):
         self._arg = arg
    
     @property
     def arg(self):
          return self._arg
```
sure its more lines of code, but that is hardly a concern when we are doing this once for a base class. The pydantic validation is pretty fancy so not sure how you would reproduce that quickly though if that is the only thing you care about.

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+Overview of EstimatorV2
+========================
+
+EstimatorV2 class estimates expectation values of quantum circuits and observables.
```
```suggestion
:class:`~EstimatorV2`estimates expectation values of quantum circuits for provided observables.
```

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -80,26 +170,29 @@
 
 from __future__ import annotations
 
+import typing
```
```suggestion
```

ikkoham committed:
Update qiskit/primitives/base/base_estimator.py

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham committed:
Use PositiveInt for shots

t-imamichi commented on qiskit/primitives/containers/data_bin.py:
```diff
@@ -0,0 +1,90 @@
+    """
+    fields = list(fields)
+    field_names = tuple(name for name, _ in fields)
+    field_types = tuple(field_type for _, field_type in fields)
```
```suggestion
    field_names, field_types = zip(*fields)
```

t-imamichi commented on qiskit/primitives/containers/data_bin.py:
```diff
@@ -0,0 +1,90 @@
+    field_types = tuple(field_type for _, field_type in fields)
+    for name in field_names:
+        if name in DataBin._RESTRICTED_NAMES:
+            raise ValueError(f"'{name}' is a restricted name for a DataBin.")
```
```suggestion
    if any(name in DataBin._RESTRICTED_NAMES for name in field_names):
        raise ValueError(f"'{name}' is a restricted name for a DataBin.")
```

ikkoham replied:
Is `name` in line 77 defined?

t-imamichi replied:
I see. `name` is not defined. The following should work, but looks a bit complicated. So, I don't recommend it.
```python
    if any((name_ := name) in DataBin._RESTRICTED_NAMES for name in field_names):
        raise ValueError(f"'{name_}' is a restricted name for a DataBin.")
```

ikkoham committed:
improve type hint

ikkoham committed:
Update qiskit/primitives/containers/data_bin.py

Co-authored-by: Takashi Imamichi <31178928+t-imamichi@users.noreply.github.com>

ihincks commented on qiskit/primitives/containers/data_bin.py:
```diff
@@ -0,0 +1,88 @@
+        return f"{type(self)}({', '.join(vals)})"
+
+
+def make_databin(
```
Very minor nit (my fault): the snake case of this function does not line up exactly with the camel case of the class. So I propose renaming it to `make_data_bin`.

ikkoham replied:
updated

chriseclectic commented on qiskit/primitives/base/base_primitive.py:
```diff
@@ -72,3 +74,38 @@ def _cross_validate_circuits_parameter_values(
+    def options(self, options: BasePrimitiveOptionsLike):
+        self._set_options(options)
+
+    def _set_options(self, options):
```
I dont think this method is necessary. You could just have something like this in the init:
```python
self._options = self._options_class()
if options:
    self._options.update(options)
```
and then any coercian has to be handled by the update method. eg it could handle a mapping , **mapping` or another options class (which isn't a mapping)

chriseclectic commented on qiskit/primitives/base/base_primitive.py:
```diff
@@ -72,3 +74,38 @@ def _cross_validate_circuits_parameter_values(
+
+    @property
+    def options(self) -> BasePrimitiveOptions:
+        """Options for BaseEstimator"""
```
```suggestion
        """Options for the primitive"""
```
This doc string will be inherited by subclasses so should leave the type generic

chriseclectic commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,44 @@ def parameters(self) -> tuple[ParameterView, ...]:
+        Returns:
+            The job object of Estimator's Result.
+        """
+        if isinstance(tasks, EstimatorTask):
```
Do we really want all this task coercion and validation in the base class or just leave it as an abstract interface? There is no reason the body of this method couldn't just be a utility function someone could call in their primitive implementation since it has no dependence on `self`. I would remove it and just have 
```python
@abstractmethod
def run(self, tasks: EstimatorTaskLike | Iterable[EstimatorTaskLike]) -> T:
```

And then an Estimator implementation could implement the body as

```
tasks = coerce_estimator_tasks(tasks)
# run logic...
```

chriseclectic commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,44 @@ def parameters(self) -> tuple[ParameterView, ...]:
+    def __init__(self, options: Optional[BasePrimitiveOptionsLike]):
+        super().__init__(options=options)
+
+    def run(self, tasks: EstimatorTaskLike | Iterable[EstimatorTaskLike]) -> T:
```
Could we remove supporting single or iterable for inputs and have either
```python
def run(self, tasks: Iterable[EstimatorTaskLike]) -> T:
```
or
```python
def run(self, *tasks: EstimatorTaskLike) -> T:
```

ihincks replied:
I'm fine with either of these. I think we should treat the RFCs as a collection of guiding ideas, rather than, necessarily, a prescription.

ikkoham replied:
Thanks. I choose iterable.

chriseclectic commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,44 @@ def parameters(self) -> tuple[ParameterView, ...]:
+        return self._run(tasks)
+
+    @abstractmethod
+    def _run(self, tasks: list[EstimatorTask]) -> T:
```
Can be removed if we make `run` the abstract method and leave validation as a utility function

ikkoham replied:
I'm fine, but your proposal is to delegate the calling of the validation to the implementor of subclasses, is that correct? I am concerned that the subclass implementer may forget to call it.

ikkoham replied:
Now `run` is the abstract method

chriseclectic reviewed (changes requested):
Review of EstimatorV2 and BasePrimitiveV2

ikkoham committed:
remove _run and make run abstractmethod

ikkoham committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>
Co-authored-by: Christopher J. Wood <cjwood@us.ibm.com>

t-imamichi commented:
How about making a directory `test/python/primitives/containers` to organize the test files?

ikkoham committed:
fix from review comments

ikkoham committed:
move test/python/primitives/containers

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

t-imamichi commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,156 @@
+        coerced_tasks = [
+            task if isinstance(task, EstimatorTask) else EstimatorTask.coerce(task)
+            for task in tasks
+        ]
```
I don't think `isinstance` check is necessary because it's done in `coerce` method.
```suggestion
        coerced_tasks = [EstimatorTask.coerce(task) for task in tasks]
```

ikkoham replied:
Right. Thanks!

t-imamichi replied:
Sorry. My suggestion is not complete because `Tuple[QuantumCircuit, ObservablesArrayLike, BindingsArrayLike]` is EstimatorTaskLike and Iterable. So, if users give a tuple, it is decomposed into each item by `for task in tasks`.

t-imamichi replied:
It might be good to extend `coerce` to accept `Iterable[EstimatorTaskLike]` perhaps.

ikkoham replied:
Your suggestion is right because users cannot pass single `EstimatorTaskLike` now (I updated the type hint from the review comments).

t-imamichi commented on qiskit/primitives/containers/options.py:
```diff
@@ -0,0 +1,42 @@
+            if not isinstance(options, BasePrimitiveOptions):
+                raise TypeError(f"Type {type(options)} is not options class")
+            for key, val in options.__dict__.items():
+                print(key, val)
```
Just in case
```suggestion
```

ikkoham committed:
Apply suggestions from code review

Co-authored-by: Takashi Imamichi <31178928+t-imamichi@users.noreply.github.com>

ikkoham committed:
update docs

ikkoham committed:
rename task to pubs

ikkoham commented on qiskit/primitives/containers/pubs_result.py:
```diff
@@ -23,10 +23,10 @@
 @frozen_dataclass
-class TaskResult:
-    """Result of task."""
+class PubsResult:
```
`PubResult` might be better?

ihincks replied:
Yes, I agree that PubResult would be better.

In general, Task <-> Pub, Tasks <-> Tasks, task <-> pub, tasks <-> pubs.

ikkoham replied:
Thanks. I updated

t-imamichi commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,29 @@ def parameters(self) -> tuple[ParameterView, ...]:
+        super().__init__(options=options)
+
+    @abstractmethod
+    def run(self, pubs: Iterable[EstimatorPubsLike]) -> Job:
```
Accoring to RFC, pubs can be not only iterable but also a single pubs.
https://github.com/Qiskit/RFCs/blob/0a5614382d6d96d66b839cf7dd9d2c165a761cf1/0015-estimator-interface.md?plain=1#L130
```python
class EstimatorBase(ABC, Generic[T]):
    ...
    def run(self, tasks: EstimatorTaskLike | Iterable[EstimatorTaskLike], **options) -> T:
        pass
```

ikkoham replied:
Thanks. It was discussed in https://github.com/Qiskit/qiskit/pull/11227#discussion_r1407955933

t-imamichi replied:
Thanks. I didn't notice the discussion.

ikkoham committed:
Pubs -> Pub

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+  to be bound to the parameters of the quantum circuits
+  (list of list of float or list of dict).
+
+The method returns a :class:`~qiskit.providers.JobV1` object, calling
```
```suggestion
Running an estimator returns a :class:`~qiskit.providers.JobV1` object, where calling
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+  (list of list of float or list of dict).
+
+The method returns a :class:`~qiskit.providers.JobV1` object, calling
+:meth:`qiskit.providers.JobV1.result()` yields the
```
I don't think sphinx knows how to remove parens?

```suggestion
the method :meth:`qiskit.providers.JobV1.result` returns
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+The method returns a :class:`~qiskit.providers.JobV1` object, calling
+:meth:`qiskit.providers.JobV1.result()` yields the
+a list of expectation values plus optional metadata like confidence intervals for
+the estimation.
```
```suggestion
corresponding expectation value estimates and metadata. 
```

ikkoham replied:
duplicated

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+Overview of EstimatorV2
+========================
+
+:class:`~BaseEstimatorV2` estimates expectation values of quantum circuits for provided observables.
```
```suggestion
:class:`~BaseEstimatorV2` is a primitive that estimates expectation values for provided quantum circuit and observable combinations.
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+
+* parameter values (:math:`\theta_k`): list of sets of values
+  to be bound to the parameters of the quantum circuits
+  (list of list of float or list of dict).
```
```suggestion
Following construction, and estimator is used by calling its :meth:`~.BaseEstimatorV2.run` method with a list of pubs (Primitive Unified Blocs).
Each pub contains three values that, together, define a computation unit of work for the estimator to complete:

* a single :class:`~qiskit.circuit.QuantumCircuit`, possibly parametrized, whose final state we define as :math:`\psi(\theta)`,

* one or more observables (specified as any :class:`~.ObservablesArrayLike`, including :class:`~.Pauli`, :class:`~.SparsePauliOp`, ``str``) that specify which expectation values to estimate, denoted :math:`H_j`, and

* a collection parameter value sets to bind the circuit against, :math:`\theta_k`.
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+The method returns a :class:`~qiskit.providers.JobV1` object, calling
+:meth:`qiskit.providers.JobV1.result()` yields the
+a list of expectation values plus optional metadata like confidence intervals for
+the estimation.
```
```suggestion
Running an estimator returns a :class:`~qiskit.providers.JobV1` object, where calling
the method :meth:`qiskit.providers.JobV1.result` results in expectation value estimates and metadata for each pub:
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,99 @@
+
+    \langle\psi(\theta_k)|H_j|\psi(\theta_k)\rangle
+
+The broadcast rule applies for observables and parameters. For more information, please check
```
```suggestion
The observables and parameter values portion of a pub can be array-valued with arbitrary dimensions, 
where standard broadcasting rules are applied, so that, in turn, the estimated result for each pub is in general array-valued as well. For more information, please check
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,29 @@ def parameters(self) -> tuple[ParameterView, ...]:
+class BaseEstimatorV2(BasePrimitiveV2):
+    """Estimator base class version 2.
+
+    An Estimator estimates expectation values of quantum circuits and observables.
```
```suggestion
    An estimator estimates expectation values for provided quantum circuit and observable combinations.
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,29 @@ def parameters(self) -> tuple[ParameterView, ...]:
+
+    @abstractmethod
+    def run(self, pubs: Iterable[EstimatorPubLike]) -> Job:
+        """Run the pubs of the estimation of expectation value(s).
```
```suggestion
        """Estimate expectation values for each provided pub (Primitive Unified Bloc).
```

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,29 @@ def parameters(self) -> tuple[ParameterView, ...]:
+                ``(QuantumCircuit, observables, parameter_values)``
+
+        Returns:
+            The job object of Estimator's Result.
```
```suggestion
            A job object that contains results.
```

mtreinish commented on requirements.txt:
```diff
@@ -9,3 +9,4 @@ python-dateutil>=2.8.0
 stevedore>=3.0.0
 typing-extensions; python_version<'3.11'
 symengine>=0.9,!=0.10.0
+pydantic
```
Why is pydantic needed here? None of the RFCs outlined a use for this in any of the base primitive classes. I'm pretty wary of adding a new dependency just for input validation on options. I don't think we should be using it on any of the base interface definitions which the options rfc said too. If implementations want to use pydantic interfaces that should be compatible but it shouldn't be a hard requirement. Also using it just for the new statevector estimator implementation is kind of overkill to enforce that shots is an int and >= 0.  In general I'm a fan of pydantic as a library (especially compared to marshmallow which Qiskit's previously used it for a similar reason). But I really don't think it's something we should be adding it as a requirement here.

As an aside, this almost certainly has a lower bound since of at least >=2.0 since I'm pretty sure you're using pydantic v2 here. So you should probably have this set at `pydantic>=2`.

ikkoham replied:
[This RFC](https://github.com/Qiskit/RFCs/blob/master/0018-primitive-options-type.md) proposed using Python and **pydantic dataclasses** as the object type of primitive options. Of course, we could use Python's built-in dataclass, but since we are determined to use pydantic in a subclass (i.e. qiskit-ibm-runtime), I don't know if mixing the two is a good decision.
However, it is understandable to be cautious about adding new dependencies.
I'm very difficult to decide which one is better. Maybe it should be optional dependency to begin with.

Thanks for the pydantic version. It is indeed true.

ihincks replied:
There are some other comments suggesting removing the use of dataclasses to help implement, eg, Pub. Since dataclasses aren't doing heavy lifting in those implementations (they give a repr), I am now in favour of making that change.

ikkoham replied:
I updated pydantic is optional.

jyu00 replied:
Sorry if it wasn't clear in the RFC, which meant to say we'd use Python dataclasses for options in the base, and implementers can choose to use more advanced dataclasses like pydantic.

ihincks commented on qiskit/primitives/base/base_primitive.py:
```diff
@@ -72,3 +74,23 @@ def _cross_validate_circuits_parameter_values(
+    version = 2
+    _options_class: type[BasePrimitiveOptions] = BasePrimitiveOptions
+
+    def __init__(self, options: Optional[BasePrimitiveOptionsLike] = None):
```
```suggestion
    def __init__(self, options: BasePrimitiveOptionsLike | None = None):
```

ihincks reviewed (commented):
Nit: could we please favour | instead of Union and Optional? I'm making this suggestion for the sake of consistency, at least in our new files. It's somewhat annoying that Python has both of these styles around, but my understanding is | is the preferred stye. (If qiskit already has a style guide that contradicts this, pls ignore me )

ikkoham committed:
make pydantic optional

ikkoham commented:
@ihincks `|` may not be used in Python 3.8 instead of `Union` in some case, e.g. type alias (`A = B | C` is invalid). Whether it is `Optional` or `A | None` can be either, and there is no fixed rule, at least in Qiskit. So I'm fine with using `A | None`.

t-imamichi commented on qiskit/primitives/containers/pub_result.py:
```diff
@@ -0,0 +1,37 @@
+
+
+@frozen_dataclass
+class PubResult:
```
Is it not necessary to set `@HAS_PYDANTIC.require_in_instance`?

ikkoham replied:
If there is not pydantic, `dataclasses.dataclass` is used instead of `pydantic.dataclass`.

ikkoham committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ikkoham committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ikkoham committed:
fix lint

ikkoham committed:
type hint

ikkoham committed:
fix type hint for python 3.8

ikkoham committed:
improve BindingsArray

ikkoham committed:
fix docs warning

ihincks commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,98 @@
+:class:`~BaseEstimatorV2` is a primitive that estimates expectation values for provided quantum
+circuit and observable combinations.
+
+Following construction, and estimator is used by calling its :meth:`~.BaseEstimatorV2.run` method
```
```suggestion
Following construction, an estimator is used by calling its :meth:`~.BaseEstimatorV2.run` method
```

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,355 @@
+        shape: ShapeInput | None = None,
+    ):
+        """
+        The ``shape`` argument does not need to be provided whenever it can unambiguously
```
This needs a single short stance to start before going into a full paragraph talking about shape. In general I worry that the class and init doc strings in this class are way too wordy for API docs and there should be a separate User Guide on this class

ikkoham replied:
I added short description.

chriseclectic commented on qiskit/primitives/containers/base_pub.py:
```diff
@@ -0,0 +1,36 @@
+    circuit: QuantumCircuit
+    """Quantum circuit object for the pubs."""
+
+    def validate(self):
```
What is the point of using pydantic here if we have to manually validate anyway? Might as well make it a regular class

ikkoham replied:
Pydantic is written in Rust, so performance is good, but well, it's a small class, so there won't be that much overhead.
I don't think we need this validation if we use pydantic.
But there are many voices against pydantic, so I'll make it as a regular class.

chriseclectic commented on qiskit/primitives/base/base_primitive.py:
```diff
@@ -72,3 +73,23 @@ def _cross_validate_circuits_parameter_values(
+BasePrimitive = BasePrimitiveV1
+
+
+class BasePrimitiveV2(ABC):
```
Wasn't base primitive v2 supposed to have an abstract run method? I don't really see the point of having a BasePrimitive class if not, might as well just have BaseEstimatorV2 and BaseSamplerV2

ikkoham replied:
I was aware that Generic base classes were opposed in the RFC. Can I implement it? @ihincks 

TBH, I like Generic, so I implemented Generic features for V1.

ihincks replied:
#11254, #11257, and #11529 together suggest to remove the need for mandatory options by making base sampler and estimator maintain a `shots` attribute and a `precision` attribute, respectively. If options are removed from the base primitive, the base primitive doesn't serve a concrete purpose, and so can be removed. There is always the possibility to add it later should a need arise, whereas the reverse is trickier.

chriseclectic commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -14,9 +14,98 @@
+    print(f"The primitive-job finished with result {job_result}")
+
+==============================
+Migration guide from V1 to V2
```
Should this be a user guide or tutorial doc, not hidden in API docs?

ikkoham replied:
OK, I removed this from this PR.

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,355 @@
+
+    def __init__(
+        self,
+        vals: ArrayLike | Iterable[ArrayLike] | None = None,
```
Idea: @ihincks Should we remove vals from bindings array and make it always need kwvals? vals adds a lot of complication and room for errors.

ikkoham replied:
I am neutral on the idea. But it requires additional costs, for example, if a parameter is generated by numpy.random, it is costly to convert it to dict.

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,355 @@
+        # TODO str will be used for internal data (_kwvals) instead of Parameter.
+        # This requires https://github.com/Qiskit/qiskit/issues/7107
+        kwvals = {
+            (p,) if isinstance(p, Parameter) else tuple(p): np.array(val, copy=False)
```
This is incorrect with the type hint of `ParameterLike` which also allows strings. It should be
```suggestion
            (p,) if isinstance(p, (Parameter, str)) else tuple(p): np.array(val, copy=False)
```
or `ParameterLike` should be changed to `Parameter` to remove strings

ikkoham replied:
Certainly the type hint is wrong at this point. However, the decision has been made to wait for the str bind and I will correct it when I get it in.

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,355 @@
+    @property
+    def kwvals(self) -> Dict[Tuple[str, ...], np.ndarray]:
+        """The keyword values of this array."""
+        return {_format_key(k): v for k, v in self._kwvals.items()}
```
This cannot be used to bind a circuit. We should be very careful with our conflation of Parameters and strings in this class if Qiskit QuantumCircuit no long supports using strings to identify parameters.

Eg.
```python
qc = QuantumCircuit(2)
qc.rx(Parameter("a"), 0)
qc.assign_parameters({"a": 2})
```
raises `CircuitError: 'Cannot bind parameters (a) not present in the circuit.'`

ikkoham replied:
This is the same as the comment above, and requires str bind.

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,355 @@
+        parameters = dict(zip(circuit.parameters, flat_vals))
+        parameters.update(
+            (param, val)
+            for params, vals in self._kwvals.items()
```
I think this will fail if BindingsArray was initialized with a string kwval instead of a Parameter, which should be supported if inputs a "ParameterLike" eg something like

```
qc = QuantumCircuit(2)
qc.rx(Parameter("a"), 0)

binds = BindingsArray(kwvals={"a": np.array([0])})
binds.bind(qc, (0,))
CircuitError: 'Cannot bind parameters (a) not present in the circuit.'
```

ikkoham replied:
It's the TODO comment.

chriseclectic commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,355 @@
+        """The non-keyword values of this array."""
+        return self._vals
+
+    def bind(self, circuit: QuantumCircuit, loc: Tuple[int, ...]) -> QuantumCircuit:
```
```suggestion
    def bind(self, circuit: QuantumCircuit, index: Tuple[int, ...]) -> QuantumCircuit:
```
Personally i prefer index to loc. When talking about ndarrays you talk about the index, not the loc. Eg `np.ndindex(shape)`.

ikkoham replied:
This point was corrected in @ihincks's comment, so please talk about which you prefer.

ikkoham commented:
Mergeable?

chriseclectic commented on qiskit/primitives/containers/data_bin.py:
```diff
@@ -0,0 +1,89 @@
+
+
+def make_data_bin(
+    fields: Iterable[Tuple[str, type]], shape: Tuple[int, ...] | None = None
```
If we are going all in on 3.9 python types we should not use List, Tuple, Dict, but the python classes themselves
```suggestion
    fields: Iterable[tuple[str, type]], shape: tuple[int, ...] | None = None
```
(and same elsewhere to be consistent)

ikkoham replied:
I updated the other parts, but forgot about this part. Thanks.

chriseclectic commented on qiskit/primitives/containers/dataclasses.py:
```diff
@@ -0,0 +1,35 @@
+
+from qiskit.utils.optionals import HAS_PYDANTIC
+
+if HAS_PYDANTIC:
```
Whats the point of making this optional? Its not a heavy dependency (afaik) so we should either just use it or not use it since having a conditional like this just gives more code paths to test and validate. If we want to do matts approach of leaving validation to subclasses then options needs to be a completely abstract (non-data) interface class.

chriseclectic commented on qiskit/primitives/containers/estimator_pub.py:
```diff
@@ -0,0 +1,94 @@
+
+@frozen_dataclass
+class EstimatorPub(BasePub, ShapedMixin):
+    """Pub (Primitive Unified Bloc) for Estimator.
```
```suggestion
    """Primitive Unified Bloc for Estimator.
```

chriseclectic commented on qiskit/primitives/containers/estimator_pub.py:
```diff
@@ -0,0 +1,94 @@
+
+    observables: ObservablesArray
+    parameter_values: BindingsArray = BindingsArray(shape=())
+    _shape: Tuple[int, ...] = ()
```
This is where dataclasses really look like a bad fit. Now you are adding hidden fields to its init methods, and setting them in post-init functions

chriseclectic commented on qiskit/primitives/containers/estimator_pub.py:
```diff
@@ -0,0 +1,94 @@
+from .shape import ShapedMixin
+
+
+@frozen_dataclass
```
```suggestion
```
This would really seem more natural as a regular class with all the methods it defines

chriseclectic commented on qiskit/primitives/containers/observables_array.py:
```diff
@@ -0,0 +1,265 @@
+    """An ND-array of :const:`.BasisObservable` for an :class:`.Estimator` primitive."""
+
+    __slots__ = ("_array", "_shape")
+    ALLOWED_BASIS: str = "IXYZ01+-lr"
```
Im not so sure this class really makes sense in a base class, since it is defining the bases that all estimator primitives are expected to measure in.

Also do our reference implementations support 01+-lr at the moment?

ikkoham replied:
That's a TODO comment, I thought it would be implemented by 1.0.0 so it's not a problem at this point

chriseclectic commented on qiskit/primitives/containers/options.py:
```diff
@@ -0,0 +1,41 @@
+class BasePrimitiveOptions(ABC):
+    """Base class of options for primitives."""
+
+    def update(self, options: BasePrimitiveOptions | None = None, **kwargs):
```
I would expect this to have a signature like `dict.update` but extended to also support `BasePrimitiveOptions` along with `Mapping` for the first arg.
```suggestion
     def update(self, options: BasePrimitiveOptions | Mapping | None = None, **kwargs):
```
so a single field can s

ikkoham replied:
Thanks. I updated.

chriseclectic commented on qiskit/primitives/containers/options.py:
```diff
@@ -0,0 +1,41 @@
+    def update(self, options: BasePrimitiveOptions | None = None, **kwargs):
+        """Update the options."""
+        if options is not None:
+            if not isinstance(options, BasePrimitiveOptions):
```
To support mappings something like
```python
            if isinstance(options, BasePrimitiveOptions):
                 options = options.__dict__
            if not isinstance(options, Mapping):
                  raise TypeError(...)
            for key val in options.items():
                ...
```

chriseclectic commented on qiskit/primitives/containers/pub_result.py:
```diff
@@ -0,0 +1,37 @@
+    from dataclasses import field as Field
+
+
+@frozen_dataclass
```
Make normal class?

chriseclectic reviewed (changes requested):
Bit more detailed review. i have a lot of worries with the current implementation of BindingsArray and its use of string parameters in context with QuantumCircuit.assign_parameters not longer working with string dicts

ihincks commented:
Unfortunately, I can't import the latest commit of this branch without pydantic in Python 3.9:
```
From qiskit/primitives/containers/dataclasses.py:35
     32 from dataclasses import dataclass
     34 mutable_dataclass = dataclass(frozen=False)
---> 35 frozen_dataclass = dataclass(frozen=True, slots=True)

TypeError: dataclass() got an unexpected keyword argument 'slots'
```

ihincks committed:
Remove `slots=True` in dataclass usage.

This argument was introduced in Python 3.10, so we cannot use it yet.
We don't expect that this has any performance impacts, it is just being
used here to make it more difficult to mutate.

ihincks reviewed (dismissed):
This PR has some minor implementation details that not everyone is completely satisfied with. However, it does have a working state vector simulator, it does pass all of the tests, and it does introduce all of the non-sampler ideas promised by the RFC. Since we would like to see these some version of this PR make it into the 1.0 pre-release, my preference is to merge this PR now and open the implementation detail changes as issues.

mtreinish commented on qiskit/primitives/base/base_estimator.py:
```diff
@@ -254,3 +346,30 @@ def parameters(self) -> tuple[ParameterView, ...]:
+BaseEstimator = BaseEstimatorV1
+
+
+@HAS_PYDANTIC.require_in_instance
```
This isn't really optional as it forces pydantic to be installed to use `BaseEstimatorV2`, if you tried to use `BaseEstimatorV2` without pydantic installed you'd get an error.

ikkoham commented:
Is there agreement in the lead (@ihincks, @jyu00, @chriseclectic, @mtreinish) that this PR should not be included in 1.0.0b1? I'm wondering if there have been any offline discussions.  

I will fix as much as possible up to Quantum Summit (PF). I hope main branch if merged else this branch will be used.

ihincks commented:
@ikkoham after some back and forth, those you mentioned decided that we'd prefer to wait until some issues are ironed out, rather than releasing it right now. We are very close.

ikkoham committed:
update from review comments

ikkoham committed:
Update qiskit/primitives/containers/estimator_pub.py

Co-authored-by: Christopher J. Wood <cjwood@us.ibm.com>

ikkoham committed:
Update qiskit/primitives/base/base_estimator.py

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham commented:
OK, I labeled `on hold` because this PR waits for https://github.com/Qiskit/qiskit/issues/7107 for complete `BindingsArray`. But any comments and suggestions are welcome.

t-imamichi commented on qiskit/primitives/containers/base_pub.py:
```diff
@@ -0,0 +1,47 @@
+
+        Args:
+            circuit: Quantum circuit object for the pubs.
+            validate: if True, the input data is validated during initizlization.
```
```suggestion
            validate: if True, the input data is validated during initialization.
```

t-imamichi commented on qiskit/primitives/containers/pub_result.py:
```diff
@@ -0,0 +1,45 @@
+        return self._data
+
+    @property
+    def metadata(self) -> {}:
```
```suggestion
    def metadata(self) -> dict:
```

t-imamichi commented on qiskit/primitives/containers/bindings_array.py:
```diff
@@ -0,0 +1,357 @@
+    BindingsArray,
+    NDArray,
+    "Mapping[Parameter, NDArray]",
+    "Sequence[NDArray]",
```
`NDArray` might be relaxed to `ArrayLike`?
```suggestion
    ArrayLike,
    "Mapping[Parameter, ArrayLike]",
    "Sequence[ArrayLike]",
```

ikkoham replied:
probably

t-imamichi commented on qiskit/primitives/base/base_primitive.py:
```diff
@@ -72,3 +73,23 @@ def _cross_validate_circuits_parameter_values(
+            self._options.update(options)
+
+    @property
+    def options(self) -> BasePrimitiveOptions:
```
There seems to be no setter of options. How can users specify the number of shots for each run? Do they need to make a new Estimator object with different options?

ikkoham replied:
The only way to do it is to set the attribute of the option like:

```python
estimator = Estimator()
estimator.run(pubs)
estimator.options.execution.shots = 1024 # new shots
estimator.run(pubs)
```

ikkoham committed:
Apply suggestions from code review

Co-authored-by: Takashi Imamichi <31178928+t-imamichi@users.noreply.github.com>

ikkoham committed:
lint

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

chriseclectic committed:
Add BaseEstimatorV2 class

Taken from 11227, but modified to
- remove options from BaseEstimatorV2
- add precision attribute to BaseEstimatorV2 and EstimatorPub
- remove BasePrimitiveV2 and BasePub

Co-Authored-By: Ikko Hamamura <ikkoham@users.noreply.github.com>
Co-Authored-By: Ian Hincks <2229105+ihincks@users.noreply.github.com>

chriseclectic committed:
Fix removal of BasePub

chriseclectic committed:
Remove precision from EstimatorPub

ihincks committed:
Add BaseEstimator._make_data_bin() static method

chriseclectic committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

chriseclectic committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

chriseclectic committed:
linting

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,161 @@
+
+
+@HAS_PYDANTIC.require_in_instance
+class Estimator(BaseEstimatorV2):
```
In light of #11524 and #11527, I recommend that this implementation:
 - call `super().__init__(0.0)` to indicate that the target precision is no error, because this is a full state simulation
 - not introduce any artificial shot noise via rng.normal
 - set the stds in the databin to the all zeros array because there is no error on the expectation value estimates
 - not have any options (corollary of first two points)

ihincks replied:
This comment was made when the base estimator branch had precision as a constructor argument. I'd argue for similar behaviour wrt to the run method.

ikkoham replied:
Now `super().__init__` does not have `precision`. `Estimator` needs to support arbitrary precisions.

chriseclectic committed:
Move precision to EstimatorPub and Estimator.run

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

chriseclectic committed:
Update EstimatorV2 run return type, fix some typos

ihincks committed:
Fix some minor problems

ihincks committed:
add tests

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham committed:
Merge branch 'primitives/estimator-v2-redux' into primitives/estimator-v2

ikkoham committed:
update

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,128 @@
+                setattr(self, key, val)
+
+        for key, val in kwargs.items():
+            setattr(self, key, val)
```
Let's remove these options and copy what is being done in #11566, by putting default precision either in the constructor, or as a class attribute.

ikkoham replied:
So can we ignore the approved RFC?

chriseclectic replied:
Yes

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,128 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
```
```suggestion
# (C) Copyright IBM 2023, 2024.
```

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,128 @@
+        self.options = options
+
+    def run(
+        self, pubs: Iterable[EstimatorPubLike], precision: float | None = None
```
```suggestion
        self, pubs: Iterable[EstimatorPubLike], *, precision: float | None = None
```

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,128 @@
+            if pub.precision is not None:
+                precision = pub.precision
+            else:
+                precision = run_precision
```
The `EstimatorPub.coerce` method takes care of this by returning a pub with the precision set, so you should just be able to refer to `pub.precision` everywhere in this loop.

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,128 @@
+        coerced_pubs = [EstimatorPub.coerce(pub) for pub in pubs]
+
+        for pub in coerced_pubs:
+            pub.validate()
```
See https://github.com/Qiskit/qiskit/pull/11566/files#r1456634947 for my reasoning.

```suggestion
        precision = precision or self.options.precision # TODO: switch away from options to class variable
        coerced_pubs = [EstimatorPub.coerce(pub, precision) for pub in pubs]
```

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,128 @@
+            evs = np.zeros_like(bc_circuits, dtype=np.float64)
+            stds = np.zeros_like(bc_circuits, dtype=np.float64)
+            for index in np.ndindex(*bc_circuits.shape):
+                bound_circuit = bc_circuits[index]
```
I think this will work and will prevent the need to have all bound circuits in memory at once.

```suggestion
                bound_circuit = parameter_values.bind(circuit, loc=index)
```

ihincks reviewed (commented):
Thank you for attending to the huge number of comments on this PR as it has evolved over the past few months. We are nearly there.

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,128 @@
+            setattr(self, key, val)
+
+
+class Estimator(BaseEstimatorV2):
```
There was a recommendation in the sampler statevector PR to rename it to `StatevectorSampler`. IF that goes through, this PR should do similar.

ikkoham replied:
My intent was two use case
```python
from qiskit.primitives import StatevectorEstimator
```
and
```python
from qiskit.primitives.statevector_estimator import Estimator, Options
```
the latter is assumed that users will be able to switch to other primitives simply by changing the import path.

But if `Options` is removed, this design will be dead, so I'm ok with the rename.

chriseclectic replied:
You can always `import StatevectorEstimator as Estimator` if you want it to be swap-able in a notebook/script

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham commented on qiskit/primitives/base/base_primitive.py:
```diff
@@ -24,7 +24,7 @@
 
 
-class BasePrimitive(ABC):
+class BasePrimitiveV1(ABC):
```
I'll revert this

ikkoham committed:
Apply suggestions from code review

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ikkoham committed:
revert BasePrimitiveV1

ikkoham committed:
Update qiskit/primitives/statevector_estimator.py

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ikkoham committed:
update

ikkoham committed:
rm base_pub.py

ikkoham committed:
options do not have precision

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,120 @@
+            options = Options()
+        elif not isinstance(options, Options):
+            options = Options(**options)
+        self.options = options
```
```suggestion
```
I don't think we need any init args for this estimator. In theory we could have a `default_precision`, but its unnecessary as the precision of the statevector estimator is always 0 because it computes expectation values exactly from the statevector.

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,120 @@
+    def run(
+        self, pubs: Iterable[EstimatorPubLike], *, precision: float | None = None
+    ) -> PrimitiveJob[PrimitiveResult[PubResult]]:
+        job: PrimitiveJob[PrimitiveResult[PubResult]] = PrimitiveJob(self._run, pubs, precision)
```
This can be refactored like was done in StatevectorSampler. Namley coercion in run method, then call `_run`, and having a `_run_pub` method for running a single pub.

For default precision you can do
```python
if precision is None:
    precision = 0
```
Since its going to be ignored by run anyway

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,120 @@
+        rng = _get_rng(self.options.seed)
+
+        results = []
+        for pub in coerced_pubs:
```
Inner block moved to `_run_pub` method

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,120 @@
+        self, pubs: Iterable[EstimatorPub], precision: float | None
+    ) -> PrimitiveResult[PubResult]:
+        precision = precision
+        coerced_pubs = [EstimatorPub.coerce(pub, precision) for pub in pubs]
```
Coercian moved to `run` so validation happens before job submission

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,120 @@
+                    standard_error = np.sqrt(precision)
+                    expectation_value = rng.normal(expectation_value, standard_error)
+                evs[index] = expectation_value
+                stds[index] = standard_error
```
```suggestion
                evs[index] = expectation_value
```
No point artificially adding noise, the precision/std is always zero for this Estimator since it computes expectation values exactly

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,120 @@
+        rng = seed
+    else:
+        rng = np.random.default_rng(seed)
+    return rng
```
```suggestion
```
Not needed

chriseclectic commented on requirements-optional.txt:
```diff
@@ -28,3 +28,4 @@ z3-solver>=4.7
 # manage to get a working install on a Mac the functionality should still work,
 # but as a convenience this file won't attempt the install itself.
 tweedledum; python_version<'3.11' and platform_system!="Darwin"
+pydantic>=2
```
```suggestion
```

chriseclectic reviewed (changes requested):
Changes to match StatevectorSampler in style

ikkoham commented:
OK, I remove Options, but shot noise is originally user's request for V1. We may get the request again, and let's discuss then.

ikkoham committed:
remove Options

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham committed:
refactoring

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,65 @@
+
+class StatevectorEstimator(BaseEstimatorV2):
+    """
+    Simple implementation of :class:`BaseEstimatorV2` with Statevector.
```
to match the statevector sampler

```suggestion
    Simple implementation of :class:`BaseEstimatorV2` with full state vector simulation.
```

ihincks committed:
Update qiskit/primitives/statevector_estimator.py

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,65 @@
+
+class StatevectorEstimator(BaseEstimatorV2):
+    """
+    Simple implementation of :class:`BaseEstimatorV2` with full state vector simulation.
```
It's not clear form base classes what observable are supported (and indeed this doesn't support everything currently in ObservableArray which defaults to including Pauli eigenstate projectors). It would be good to include in the docs the supported observables (eg Pauli and SparsePauliOp).

This should also include a reference to the `:class:`.Statevector` from the Quantum info module to say it does the simulation using that class.

chriseclectic commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,65 @@
+        self, pubs: Iterable[EstimatorPubLike], *, precision: float | None = None
+    ) -> PrimitiveJob[PrimitiveResult[PubResult]]:
+        if precision is not None and precision > 0:
+            raise ValueError("precision must be None or 0 for StatevectorEstimator.")
```
```suggestion
        if precision is None:
            precision = 0
```
We don't want to error if a user puts in `precision > 0` value for precision (which is a target precision), since we will always exceed it since precision is guaranteed to be 0 by implementation. This would prevent actually being able to use this in any application / algorithm workflow which uses a sensible initial guess for precision which would always be > 0.

ihincks replied:
@chriseclectic do you propose to:
1) ignore the precision
2) artificially add imprecision, as was previously implemented

The argument for (1) is that this is a statevector simulation, so it should always be perfect.
The argument for (2) is that some folks may want to use this implementation to test their ideas, which might include testing that their ideas are robust to finite noise.

chriseclectic replied:
1: in the base class we say that it is a target precision, which i implicitly assumed meant "I want you to return me a mean estimate with standard error <= this value". Maybe we can make wording more explicit.

To add to my lack of documentation comments, this classes doc string should explain how the simulation is done and say that it will always return standard error as zero because it returns exact expectation value means computed from the full state vector.

ihincks replied:
But (1) and (2) both satisfy "return me an estimate with standard error <= this value". Sure, (1) does a better estimation job, but 2 technically satisfies the requirements if the sampling is done right.

chriseclectic replied:
Sure, but neither raises an error and makes the estimator unusable in portable workflow like this does

chriseclectic replied:
Also i dont see the point in artificially sampling noise into a perfect estimate for a reference implementation, which is what this is. You can always do that from the results if you really want (and wrap it as another estimator if you like if its for testing for example)

ikkoham replied:
For users who set precision>0 expecting to run simulations with shot noise, it is difficult to understand why the implementation automatically sets it to 0 and they will not even notice that they are being automatically changed. It would be unfriendly not to warn, if not to make an error. You say there are no such users, but they exist, and they are using this feature.

ihincks replied:
I agree with @ikkoham : the main point of the `StatevectorEstimator` is not really to compute expectation values for users, it is to provide a simple reference implementation, and a tool for testing. This is because there are other implementations that are strictly better at computing expectation values: for non-exponential (albeit nisqy) runtime you need to use a runtime backend, and for more performant classical simulations (including Clifford) you can use aer. I realize this is a reversal of some previous comments I made, but I've had a bit more time to think about it now.

So, I am in favour of `(2) artificially add imprecision, as was previously implemented` because `some folks may want to use this implementation to test their ideas, which might include testing that their ideas are robust to finite noise`. If some user is in the unlikely situation that they want to call `some_application_funcion(estimator: BaseEstimatorV2)` with noiseless estimator, and they know that `some_application_function` internally hard-codes specific precisions >0, they can subclass `StatevectorEstimator`:

```python
class NoiselessEstimator(StatevectorEstimator):
    def run(self, pubs, precision = None):
        return super().run(pubs, 0.0)

result = some_application_function(NoiselessEstimator())
```

I do think that the default precision for `StatevectorEstimator` should be 0, however.

chriseclectic replied:
I don't see why a user should expect this estimator to be noisy unless we tell them that it is. I think we should explicitly state this is an ideal estimator and the precision argument is ignored since it always perfectly estimates the observable in the class documentation. 

You need to consider how an estimator is used in computational workflows. If someone is writing an algorithm (say VQE for example) that takes an estimator to run on,  it will choose some initial guess for a precision, and then perhaps depending on the variance of the result may modify this processing for subsequent evaluations. If you have some special test Estimator that only works with an argument of precision 0 this will error in these sort of applications. You don't loose anything by return a precision less than requested. 

If you want to add noise to this estimator so precision effects the results artificially you can do so (or make a second subclass estimator that adds noise), but i will not approve this PR while it raises an error or warning under a canonical workflow.

ihincks replied:
(Also, to be clear, and I think we all three agree on this point: the `StatevectorEstimator` shouldn't raise errors when given `precision>0` in any case. Thanks @chriseclectic for flagging this, I shouldn't have missed that. That would make the thing non-portable, in addition to being annoying. So I'm in particular trying to get to the bottom of choosing between 1 and 2, outlined near the top of this thread)

chriseclectic replied:
If you want the class to add noise for precision > 0 I am fine with that, just dont add a shots option, all you shuold need is a `seed` field (like sampler) and initialize an rng in run and do something like `rng.normal(evs, precision)` (im not exactly sure correct syntax for normal distribution sampling in numpy)

ihincks replied:
Great, let's go with (2) and implement rng seeding in the same way as the `StatevectorSampler`. No need to mention shots anywhere.

chriseclectic reviewed (changes requested):
The raises on precision > 0 needs to be removed. Other than that I would suggest adding more detailed API documentation to this class rather than referring to BaseEstimatorV2's documentation since it is more a developer facing class, not a user facing class (I am fine with the docs being done as follow up to get this into the release candidate though).

t-imamichi commented:
As class name becomes `StatevectorEstimator`, it would be nice to rename the test file name with `test_statevector_estimator.py`.

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ikkoham committed:
add seed

t-imamichi commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -46,9 +69,11 @@ def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
         return PrimitiveResult([self._run_pub(pub) for pub in pubs])
 
     def _run_pub(self, pub: EstimatorPub) -> PubResult:
+        rng = _get_rng(self._seed)
```
You can simplify it (Chris' comment https://github.com/Qiskit/qiskit/pull/11566#discussion_r1455813113)
```suggestion
        rng = np.random.default_rng(self._seed)
```

ikkoham replied:
Cool! Thanks

ikkoham committed:
Update qiskit/primitives/statevector_estimator.py

Co-authored-by: Takashi Imamichi <31178928+t-imamichi@users.noreply.github.com>

ikkoham committed:
clean

t-imamichi commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,93 @@
+            obs = SparsePauliOp(paulis, coeffs)  # TODO: support non Pauli operators
+            expectation_value = final_state.expectation_value(obs)
+            if precision != 0:
+                expectation_value = rng.normal(expectation_value, precision)
```
You need to make sure `expectation_value` is float to use `normal`. Otherwise, it raises an error as follows.
```python
In [4]: rng.normal(1j, 0.1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 rng.normal(1j, 0.1)

File numpy/random/_generator.pyx:1220, in numpy.random._generator.Generator.normal()

File _common.pyx:585, in numpy.random._common.cont()

TypeError: float() argument must be a string or a real number, not 'complex'
```

ikkoham replied:
Thanks. You are right. Actually, I haven't validate this since V1.

ikkoham committed:
validate non hermitian

t-imamichi reviewed (dismissed):
LGTM

ikkoham committed:
Merge branch 'main' into primitives/estimator-v2

ihincks commented on qiskit/primitives/statevector_estimator.py:
```diff
@@ -0,0 +1,95 @@
+
+class StatevectorEstimator(BaseEstimatorV2):
+    """
+    Simple implementation of :class:`BaseEstimatorV2` with full state vector simulation.
```
Please keep in mind that the reno PR will add an example to this docstring. But in addition to that, we could add this, or anything else that you suggest:

```suggestion
    Simple implementation of :class:`BaseEstimatorV2` with full state vector simulation.
    
    This class is implemented via :class:`~.Statevector` which turns provided circuits into 
    pure state vectors. These states are subsequently acted on by :class:~.SparsePauliOp`,
    which implies that, at present, this implementation is only compatible with Pauli-based
    observables.
```

ikkoham replied:
Thanks, TBH, I didn't improve the documentation since your reno PR will improve it. Sorry for my lazy work.

ihincks replied:
No, that's okay, I expected the change to happen in the reno, not here. Just trying to get through the comments so we can merge.

ihincks reviewed (dismissed):
This PR LGTM, but I'm going to wait for @chriseclectic make any final comments about the docstring or other lingering concerns. My preference is to get this merged (overdue) and make final tweaks as needed.

ikkoham committed:
Update qiskit/primitives/statevector_estimator.py

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ikkoham committed:
black

ihincks reviewed (approved):
LGTM

chriseclectic reviewed (approved):
LGTM
